services/stats/memstats/stats.go

```diff
@@ -0,0 +1,181 @@
+package memstats
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/rudderlabs/rudder-server/services/stats"
+)
+
+var _ stats.Stats = (*Store)(nil)
+
+var _ stats.Measurement = (*Measurement)(nil)
+
+type Store struct {
```

Comment:
all fields in the `Store` struct could be unexported and intialization could be taken care once and for all by a producer function using `Opts` if necessary. This would eliminate the need for the `init` method and the `sync.Once` field e.g.

```go
type Opt func(*Store)

var WithNow = func(now func() time.Time) Opt {
	return func(s *Store) {
		s.now = now
	}
}

// New returns a new stats.Stats implementation that stores stats in-memory.
func New(opts ...Opt) *Store {
	s := &Store{
		byKey: make(map[string]*measurement),
		now:   time.Now,
	}
	for _, opt := range opts {
		opt(s)
	}
	return s
}

type Store struct {
	mu    sync.Mutex
	byKey map[string]*measurement
	now func() time.Time
}
```

and other packages could simply do
```go
store := memstats.New(
		memstats.WithNow(func() time.Time {
			return now
		}))
```

Reply:
Your suggestion and my implementation are good patterns to initialise components in Go. I use and like both depending on the situation.

I don't see it as a problem having unexported fields for options; keep in mind most standard libraries are using this pattern.

The downside of the lazy init approach is that we have to call `.init` in every exported method. The `with` approach also has an overhead of as well but is safer indeed.